src/headerssync.cpp

```diff
@@ -0,0 +1,318 @@
+#include <headerssync.h>
+#include <logging.h>
+#include <util/check.h>
+#include <pow.h>
+#include <timedata.h>
+
+// Store a commitment to a header every HEADER_COMMITMENT_FREQUENCY blocks
+constexpr size_t HEADER_COMMITMENT_FREQUENCY = 571;
+// Pull out headers for acceptance during redownload once our buffer exceeds this amount
+constexpr size_t REDOWNLOAD_HEADERS_THRESHOLD = 25*HEADER_COMMITMENT_FREQUENCY;
+// Ensure that the redownload buffer is a multiple of the header commitment
+// frequency, so that we aren't wasting memory.
+static_assert(REDOWNLOAD_HEADERS_THRESHOLD % HEADER_COMMITMENT_FREQUENCY == 0);
+
+// Our memory analysis assumes 48 bytes for a CompressedHeader (so we should
+// re-calculate parameters if we compress further)
+static_assert(sizeof(CompressedHeader) == 48);
+
+HeadersSyncState::HeadersSyncState(NodeId id, const Consensus::Params& consensus_params) :
+    m_id(id), m_consensus_params(consensus_params)
+{ }
+
+// Free any memory in use, and mark this object as no longer usable.
+void HeadersSyncState::Finalize()
+{
+    Assume(m_download_state != State::FINAL);
+    m_header_commitments.clear();
+    m_last_header_received.SetNull();
+    m_blockhash_with_sufficient_work.SetNull();
+    std::deque<CompressedHeader>().swap(m_redownloaded_headers);
+    m_redownload_buffer_last_hash.SetNull();
+    m_redownload_buffer_first_prev_hash.SetNull();
+    m_process_all_remaining_headers = false;
+    m_current_height = 0;
+
+    m_download_state = State::FINAL;
+}
+
+// Initialize the parameters for this headers download, validate this first
+// batch, and request more headers.
+std::optional<CBlockLocator> HeadersSyncState::StartInitialDownload(const CBlockIndex* chain_start,
+        const std::vector<CBlockHeader>& initial_headers, const arith_uint256&
+        minimum_required_work, CBlockLocator&& chain_start_locator)
+{
+    // A new instance of this object should be instantiated for every headers
+    // sync, so that we don't reuse our salted hasher between syncs.
+    assert(m_download_state == State::INITIAL_DOWNLOAD);
+
+    m_chain_start = chain_start;
+    m_minimum_required_work = minimum_required_work;
+    m_current_chain_work = chain_start->nChainWork;
+    m_current_height = chain_start->nHeight;
+    m_chain_start_locator = std::move(chain_start_locator);
+
+    // Overestimate the number of blocks that could possibly exist on this
+    // chain using 6 blocks/second (fastest blockrate given the MTP rule) times
+    // the number of seconds from the last allowed block until today. This
+    // serves as a memory bound on how many commitments we might store from
+    // this peer, and we can safely give up syncing if the peer exceeds this
+    // bound, because it's not possible for a consensus-valid chain to be
+    // longer than this.
```

Comment:
In commit "Utilize anti-DoS headers download strategy"

Arguably, it's not exactly true that no consensus-valid chain can be longer, but it's not possible for the peer to have such a chain at the time the sync starts (it'd need a tip timestamp that is in the future at the time the sync starts).

Reply:
Reworked the comment to try to make that clearer, let me know if it looks better now...

Reply:
Looks good.